agent/tlsfwd/tlsfwd.go

```diff
@@ -0,0 +1,201 @@
+// Package tlsfwd implements a simple TLS forwarder
+package tlsfwd
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"net/netip"
+	"strconv"
+	"sync/atomic"
+
+	"github.com/darvaza-proxy/core"
+	"github.com/darvaza-proxy/darvaza/shared/proxy"
+	"github.com/darvaza-proxy/darvaza/shared/tls/sni"
+	"github.com/darvaza-proxy/middleware"
+)
+
+// Server is a simple TLS Frowarder
+type Server struct {
+	ps        []string
+	ds        []*sni.Dispatcher
+	res       net.IP
+	redir     bool
+	wg        core.WaitGroup
+	ctx       context.Context
+	cancel    context.CancelFunc
+	cancelled atomic.Bool
+	err       atomic.Value
+}
+
+// NewTLSFwd constructs and returns a new TLS forwarder from arguments
+func NewTLSFwd(ports []int, options ...func(*Server) error) (*Server, error) {
+	ctx1, cancel := context.WithCancel(context.Background())
+
+	pts := make([]string, 0)
+	for _, p := range ports {
+		port := ":" + strconv.Itoa(p)
+		pts = append(pts, port)
+	}
+
+	srv := Server{
+		ps:     pts,
+		ctx:    ctx1,
+		cancel: cancel,
+	}
+
+	for _, opt := range options {
+		err := opt(&srv)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &srv, nil
+}
+
+// SetRedir will allow http -> https redirection from port 80 to 443
+func SetRedir(s *Server) error {
+	s.redir = true
+	return nil
+}
+
+func getDefaultTLSHandler(c *tls.ClientHelloInfo) sni.Handler {
+	return func(ctx context.Context,
+		conn net.Conn,
+	) error {
+		return tlsHandler(ctx, c, conn)
+	}
+}
+
+func tlsHandler(ctx context.Context, c *tls.ClientHelloInfo,
+	conn net.Conn,
+) error {
+	defer conn.Close()
+
+	select {
+	case <-ctx.Done():
+		return nil
+	default:
+	}
+
+	upstream, err := addrFromName(c.ServerName)
+	if err != nil {
+		return err
+	}
+
+	errm := proxy.Forward(ctx, conn, upstream)
+	fmt.Println(errm)
+	return errm
+}
+
+func addrFromName(s string) (netip.AddrPort, error) {
+	upIPs, err := net.LookupIP(s)
+	if err != nil {
+		fmt.Println(err)
+		return netip.AddrPort{}, err
+	}
+
+	addrToUse, ok := netip.AddrFromSlice(upIPs[0])
+	if !ok {
+		return netip.AddrPort{}, err
+	}
+
+	upstream := netip.AddrPortFrom(addrToUse, 443)
+
+	return upstream, nil
+}
+
+// Serve will start serving connections caught by a Server
+func (s *Server) Serve() error {
+	var ok bool
+
+	defer func() {
+		if !ok {
+			for _, k := range s.ds {
+				_ = k.Close
+			}
+		}
+	}()
+
+	errHTTP := s.prepareRedir()
+	if errHTTP != nil {
+		return errHTTP
+	}
+
+	err := s.prepareDispatchers()
+	if err != nil {
+		return err
+	}
+	ok = true
+	return s.wg.Wait()
+}
+
+func (s *Server) prepareRedir() error {
+	if s.redir {
+		s.wg.Go(func() error {
+			return http.ListenAndServe(":80",
+				middleware.NewHTTPSRedirectHandler(0))
+		})
+	}
+	return nil
+}
+
+func (s *Server) prepareDispatchers() error {
+	for _, p := range s.ps {
+		dispatcher := &sni.Dispatcher{
+			GetHandler: getDefaultTLSHandler,
+		}
+		s.ds = append(s.ds, dispatcher)
+		ll, err := net.Listen("tcp", p)
```

Comment:
for better error handling all listeners should be opened before spawning the workers

Reply:
pushed